File: Stats.Delphi/Basic_Wizard.pas
unit Basic_Wizard;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, ExtCtrls, ComCtrls;

type
  TfrmBasicWizard = class(TForm)
    Panel1: TPanel;
    btnHelp: TButton;
    btnBack: TButton;
    btnNext: TButton;
    btnCancel: TButton;
    Panel2: TPanel;
    lblHeader: TLabel;
    PageControl: TPageControl;
    procedure btnCancelClick(Sender: TObject);
    procedure FormShow(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure btnNextClick(Sender: TObject);
    procedure btnBackClick(Sender: TObject);
  private
    { Private declarations }
    DefTabWidth : Integer;
    FFormatString: string;
    procedure SetFormatString(const Value: string);
  protected
    moveback: boolean;
    moveforward: boolean;
    chartindex: Integer;
    procedure RefreshWizardControls; virtual;
    procedure DoMoveForward; virtual;
    procedure DoMoveBack; virtual;
  public
    { Public declarations }
    procedure SetupTabs(Const NumTabs: Integer; RichEdit: TRichEdit);
    property FormatString: string read FFormatString write SetFormatString;
  end;

var
  frmBasicWizard: TfrmBasicWizard;

implementation

Uses Math387;

{$R *.DFM}

procedure TfrmBasicWizard.btnCancelClick(Sender: TObject);
begin
  Self.Close;
end;

procedure TfrmBasicWizard.SetupTabs(const NumTabs: Integer;
  RichEdit: TRichEdit);
var i : Integer;
begin
  RichEdit.Paragraph.Tab[0] := 0;
  for i := 1 to RichEdit.Paragraph.TabCount - 1 do
    RichEdit.Paragraph.Tab[i] := RichEdit.Paragraph.Tab[i-1]+DefTabWidth;
end;

procedure TfrmBasicWizard.FormShow(Sender: TObject);
begin
  { improve this }
  DefTabWidth := Max(60,Canvas.TextWidth('+'+FormatString)+10);
  RefreshWizardControls;
end;

procedure TfrmBasicWizard.SetFormatString(const Value: string);
begin
  FFormatString := Value;
end;

procedure TfrmBasicWizard.FormCreate(Sender: TObject);
begin
  FFormatString := '0.0000';
  moveback := true;
  moveforward := true;
end;

procedure TfrmBasicWizard.btnNextClick(Sender: TObject);
begin
  DoMoveForward;
  RefreshWizardControls;
end;

procedure TfrmBasicWizard.btnBackClick(Sender: TObject);
begin
  DoMoveBack;
  RefreshWizardControls;
end;

// Refresh wizard controls
procedure TfrmBasicWizard.RefreshWizardControls;
begin
  btnBack.Enabled := PageControl.ActivePage.PageIndex > 0;
  btnNext.Enabled := PageControl.ActivePage.PageIndex < PageControl.PageCount-1;
  lblHeader.Caption := PageControl.ActivePage.Caption;
end;

// Called when Back button is clicked
procedure TfrmBasicWizard.DoMoveBack;
var index: Integer;
begin
  index := PageControl.ActivePageIndex;
  if (index>0) and (moveback) then
  begin
    Dec(index);
    PageControl.ActivePageIndex := index;
  end;
end;

// Called when Next button is clicked
procedure TfrmBasicWizard.DoMoveForward;
var index: Integer;
begin
  index := PageControl.ActivePageIndex;
  if (index<PageControl.PageCount) and (moveforward) then
  begin
    Inc(index);
    PageControl.ActivePageIndex := index;
  end;
end;

end.

File: Stats.CPP/Basic_Wizard.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Basic_Wizard.h"
#include "Math387.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmBasicWizard *frmBasicWizard;
//---------------------------------------------------------------------------
__fastcall TfrmBasicWizard::TfrmBasicWizard(TComponent* Owner)
	: TForm(Owner)
{
	FFormatString = "0.0000";
  movebackward = true;
  moveforward = true;
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicWizard::btnCancelClick(TObject *Sender)
{
	Close();
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicWizard::RefreshWizardControls()
{
  btnBack->Enabled = (PageControl->ActivePage->PageIndex > 0);
  btnNext->Enabled = (PageControl->ActivePage->PageIndex < PageControl->PageCount-1);
  lblHeader->Caption = PageControl->ActivePage->Caption;
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicWizard::DoMoveForward()
{
  int index = PageControl->ActivePageIndex;
  if ((index<PageControl->PageCount) && (moveforward))
  {
      index ++;
    	PageControl->ActivePageIndex = index;
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicWizard::DoMoveBackward()
{
  int index = PageControl->ActivePageIndex;
  if ((index>0) && (movebackward))
  {
      index --;
    	PageControl->ActivePageIndex = index;
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicWizard::SetupTabs(int NumTabs, TRichEdit *RichEdit)
{
	if (DefTabWidth == 0 && Visible) {
		FormShow(this);
	}
	RichEdit->Paragraph->Tab[0] = 0;
	for (int i = 1;i< RichEdit->Paragraph->TabCount;i++)
	{
		RichEdit->Paragraph->Tab[i] = RichEdit->Paragraph->Tab[i-1]+DefTabWidth;
	}
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicWizard::FormShow(TObject *Sender)
{
	DefTabWidth = Max(70,Canvas->TextWidth("+"+FormatString)+10);
  RefreshWizardControls();
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicWizard::SetFormatString(AnsiString Value)
{
	FFormatString = Value;
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicWizard::btnNextClick(TObject *Sender)
{
	DoMoveForward();
  RefreshWizardControls();
}
//---------------------------------------------------------------------------
void __fastcall TfrmBasicWizard::btnBackClick(TObject *Sender)
{
		DoMoveBackward();
    RefreshWizardControls();
}
//---------------------------------------------------------------------------

